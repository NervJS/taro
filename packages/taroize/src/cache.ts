import { Wxml } from './wxml'
import { cloneDeep } from 'lodash'
const cacheMap = new Map<string, Wxml>()

export function getCacheWxml (dirpath: string): Wxml | undefined {
  return cloneDeep(cacheMap.get(dirpath))
}

export function saveCacheWxml (dirpath: string, wxml: Wxml) {
  cacheMap.set(dirpath, cloneDeep(wxml))
}
