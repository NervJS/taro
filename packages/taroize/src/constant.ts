export const reserveKeyWords = new Set([
  'break',
  'case',
  'catch',
  'continue',
  'default',
  'delete',
  'do',
  'else',
  'finally',
  'for',
  'function',
  'if',
  'in',
  'instanceof',
  'new',
  'return',
  'switch',
  'this',
  'throw',
  'try',
  'typeof',
  'var',
  'void',
  'while',
  'with',
  'abstract',
  'boolean',
  'byte',
  'char',
  'class',
  'const',
  'debugger',
  'double',
  'enum',
  'export',
  'extends',
  'final',
  'float',
  'goto',
  'implements',
  'import',
  'int',
  'interface',
  'long',
  'native',
  'package',
  'private',
  'protected',
  'public',
  'short',
  'static',
  'super',
  'synchronized',
  'throws',
  'transient'
])
