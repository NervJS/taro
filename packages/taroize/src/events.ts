export const specialEvents = new Map<string, string>()
specialEvents.set('bindtimeupdate', 'onTimeUpdate')
specialEvents.set('bindgetphoneNumber', 'onGetPhoneNumber')
specialEvents.set('bindgetrealnameauthinfo', 'onGetRealnameAuthInfo')
specialEvents.set('bindopensetting', 'onOpenSetting')
specialEvents.set('bindscancode', 'onScanCode')
specialEvents.set('bindstatechange', 'onStateChange')
specialEvents.set('bindhtouchmove', 'onHTouchMove')
specialEvents.set('bindvtouchmove', 'onVTouchMove')
specialEvents.set('bindcolumnchange', 'onColumnChange')
specialEvents.set('bindscrolltoupper', 'onScrollToUpper')
specialEvents.set('bindscrolltolower', 'onScrollToLower')
specialEvents.set('bindanimationfinish', 'onAnimationFinish')
specialEvents.set('bindfullscreenchange', 'onFullscreenChange')
specialEvents.set('bindtouchstart', 'onTouchStart')
specialEvents.set('bindtouchmove', 'onTouchMove')
specialEvents.set('bindtouchcancel', 'onTouchCancel')
specialEvents.set('bindtouchend', 'onTouchEnd')
specialEvents.set('bindlongpress', 'onLongPress')
specialEvents.set('bindlongclick', 'onLongClick')
specialEvents.set('bindtransitionend', 'onTransitionEnd')
specialEvents.set('bindanimationstart', 'onAnimationStart')
specialEvents.set('bindanimationtteration', 'onAnimationIteration')
specialEvents.set('bindanimationend', 'onAnimationEnd')
specialEvents.set('bindtouchforcechange', 'onTouchForceChange')
specialEvents.set('bindtap', 'onClick')
specialEvents.forEach((value, key) => {
  specialEvents.set(key.replace(/^bind/, 'catch'), value)
})
