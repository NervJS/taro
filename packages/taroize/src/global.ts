export const usedComponents = new Set<string>()

export const errors: string[] = []
