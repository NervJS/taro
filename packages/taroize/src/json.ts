import { buildTemplate } from './utils'
import * as t from 'babel-types'

export function parseJSON (json?: string) {
  if (!json) {
    return
  }
  return buildTemplate(`(${json})`) as t.ObjectExpression
}
