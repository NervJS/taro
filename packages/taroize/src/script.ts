import * as t from 'babel-types'
import traverse, { NodePath, Visitor } from 'babel-traverse'
import * as template from 'babel-template'
import { buildImportStatement, codeFrameError, buildRender, buildBlockElement, parseCode } from './utils'
import { WXS } from './wxml'
import { PageLifecycle, Lifecycle } from './lifecycle'
import { usedComponents } from './global'

const buildDecorator = (type: string) => t.decorator(
  t.callExpression(t.identifier('withWeapp'), [t.stringLiteral(type)])
)

export function parseScript (
  script?: string,
  returned?: t.Expression,
  json?: t.ObjectExpression,
  wxses: WXS[] = [],
  refId?: Set<string>
) {
  script = script || 'Page({})'
  if (t.isJSXText(returned as any)) {
    const block = buildBlockElement()
    block.children = [returned as any]
    returned = block
  }
  let ast = parseCode(script)
  let classDecl!: t.ClassDeclaration
  let foundWXInstance = false
  const vistor: Visitor = {
    BlockStatement (path) {
      path.scope.rename('wx', 'Taro')
    },
    CallExpression (path) {
      const callee = path.get('callee')
      if (callee.isIdentifier()) {
        const name = callee.node.name
        if (name === 'getApp' || name === 'getCurrentPages') {
          callee.replaceWith(
            t.memberExpression(t.identifier('Taro'), callee.node)
          )
        }
      }
      if (callee.isMemberExpression()) {
        const object = callee.get('object')
        if (object.isIdentifier({ name: 'wx' })) {
          object.replaceWith(t.identifier('Taro'))
        }
      }
      if (
        callee.isIdentifier({ name: 'Page' }) ||
        callee.isIdentifier({ name: 'Component' }) ||
        callee.isIdentifier({ name: 'App' })
      ) {
        foundWXInstance = true
        const componentType = callee.node.name
        classDecl = parsePage(
          path,
          returned || t.nullLiteral(),
          json,
          componentType,
          refId
        )!
        if (componentType !== 'App') {
          classDecl.decorators = [buildDecorator(componentType)]
        }
        ast.program.body.push(
          classDecl,
          t.exportDefaultDeclaration(t.identifier(componentType !== 'App' ? defaultClassName : 'App'))
        )
        // path.insertAfter(t.exportDefaultDeclaration(t.identifier(defaultClassName)))
        path.remove()
      }
    }
  }

  traverse(ast, vistor)

  if (!foundWXInstance) {
    ast = parseCode(script + ';Component({})')
    traverse(ast, vistor)
  }

  const taroComponentsImport = buildImportStatement('@tarojs/components', [
    ...usedComponents
  ])
  const taroImport = buildImportStatement('@tarojs/taro', [], 'Taro')
  const withWeappImport = buildImportStatement(
    '@tarojs/with-weapp',
    [],
    'withWeapp'
  )
  ast.program.body.unshift(
    taroComponentsImport,
    taroImport,
    withWeappImport,
    ...wxses.filter(wxs => !wxs.src.startsWith('./wxs__')).map(wxs => buildImportStatement(wxs.src, [], wxs.module))
  )

  return ast
}

const defaultClassName = '_C'

const staticProps = ['externalClasses', 'relations', 'options']

function parsePage (
  pagePath: NodePath<t.CallExpression>,
  returned: t.Expression,
  json?: t.ObjectExpression,
  componentType?: string,
  refId?: Set<string>
) {
  const stateKeys: string[] = []
  let methods: NodePath<
    t.ObjectProperty | t.ObjectMethod
  >[] = []
  pagePath.traverse({
    CallExpression (path) {
      const callee = path.get('callee')
      if (callee.isIdentifier()) {
        const name = callee.node.name
        if (name === 'getApp' || name === 'getCurrentPages') {
          callee.replaceWith(
            t.memberExpression(t.identifier('Taro'), callee.node)
          )
        }
      }
      if (callee.isMemberExpression()) {
        const object = callee.get('object')
        if (object.isIdentifier()) {
          const methodName = object.node.name
          const hooks = ['onLoad', 'onShow', 'onReady', 'onHide', 'onUnload', 'onError', 'onLaunch']
          hooks.forEach(hook => {
            if (methodName === hook) {
              object.replaceWith(t.identifier(PageLifecycle.get(methodName)!))
            }
          })
          if (methodName === 'wx') {
            object.replaceWith(t.identifier('Taro'))
          }
        }
      }
    },
    ObjectProperty (path) {
      const { key, value } = path.node
      if (!t.isIdentifier(key, { name: 'methods' }) || path.parentPath !== pagePath.get('arguments')[0] || !t.isObjectExpression(value)) {
        return
      }
      methods = path.get('value.properties') as NodePath<
        t.ObjectProperty | t.ObjectMethod
      >[]
      path.remove()
    }
  })
  if (refId) {
    refId.forEach(id => {
      if (!stateKeys.includes(id)) {
        stateKeys.push(id)
      }
    })
  }
  const propsKeys: string[] = []
  const arg = pagePath.get('arguments')[0]
  if (!arg || !arg.isObjectExpression()) {
    throw codeFrameError(arg.node, `${componentType || '组件'} 的第一个参数必须是一个对象才能转换。`)
  }
  const defaultProps: { name: string, value: any }[] = []
  const props = arg.get('properties')
  const properties = props.filter(p => !p.isSpreadProperty()).concat(methods) as NodePath<
    t.ObjectProperty | t.ObjectMethod
  >[]

  let classBody = properties.map(prop => {
    const key = prop.get('key')
    const value = prop.get('value')
    const params = prop.isObjectMethod()
      ? prop.node.params
      : value.isFunctionExpression() || value.isArrowFunctionExpression()
        ? value.node.params
        : []
    const isAsync = prop.isObjectMethod()
      ? prop.node.async
      : value.isFunctionExpression() || value.isArrowFunctionExpression()
        ? value.node.async
        : false
    if (!key.isIdentifier()) {
      throw codeFrameError(key.node, 'Page 对象的键值只能是字符串')
    }
    const name = key.node.name
    const currentStateKeys: string[] = []
    if (name === 'data') {
      if (value.isObjectExpression()) {
        value
          .get('properties')
          .map(p => p.node)
          .forEach(prop => {
            if (t.isObjectProperty(prop)) {
              if (t.isStringLiteral(prop.key)) {
                currentStateKeys.push(prop.key.value)
              }
              if (t.isIdentifier(prop.key)) {
                currentStateKeys.push(prop.key.name)
              }
            }
          })
      }
      return t.classProperty(t.identifier('state'), value.node)
    }
    if (name === 'properties') {
      const observeProps: { name: string, observer: any }[] = []
      if (value.isObjectExpression()) {
        value
          .get('properties')
          .map(p => p.node)
          .forEach(prop => {
            if (t.isObjectProperty(prop)) {
              let propKey: string | null = null
              if (t.isStringLiteral(prop.key)) {
                propKey = prop.key.value
              }
              if (t.isIdentifier(prop.key)) {
                propKey = prop.key.name
                // propsKeys.push(prop.key.name)
              }
              if (t.isObjectExpression(prop.value) && propKey) {
                for (const p of prop.value.properties) {
                  if (t.isObjectProperty(p)) {
                    let key: string | null = null
                    if (t.isStringLiteral(p.key)) {
                      key = p.key.value
                    }
                    if (t.isIdentifier(p.key)) {
                      key = p.key.name
                    }
                    if (key === 'value') {
                      defaultProps.push({
                        name: propKey,
                        value: p.value
                      })
                    } else if (key === 'observer') {
                      observeProps.push({
                        name: propKey,
                        observer: p.value
                      })
                    }
                  }
                  if (t.isObjectMethod(p) && t.isIdentifier(p.key, { name: 'observer' })) {
                    observeProps.push({
                      name: propKey,
                      observer: t.arrowFunctionExpression(p.params, p.body, p.async)
                    })
                  }
                }
              }
              if (propKey) {
                propsKeys.push(propKey)
              }
            }
          })
      }
      currentStateKeys.forEach(s => {
        if (propsKeys.includes(s)) {
          throw new Error(`当前 Component 定义了重复的 data 和 properites: ${s}`)
        }
      })
      stateKeys.push(...currentStateKeys)
      return t.classProperty(t.identifier('_observeProps'), t.arrayExpression(
        observeProps.map(p => t.objectExpression([
          t.objectProperty(
            t.identifier('name'),
            t.stringLiteral(p.name)
          ),
          t.objectProperty(
            t.identifier('observer'),
            p.observer
          )
        ]))
      ))
    }
    if (PageLifecycle.has(name)) {
      const lifecycle = PageLifecycle.get(name)!
      const node = value.node as
        | t.FunctionExpression
        | t.ArrowFunctionExpression
      const method = t.classMethod(
        'method',
        t.identifier(lifecycle),
        params,
        node ? node.body as t.BlockStatement : (prop.get('body') as any).node
      )
      method.async = isAsync
      return method
    }
    if (prop.isObjectMethod()) {
      const body = prop.get('body')
      return t.classProperty(
        t.identifier(name),
        t.arrowFunctionExpression(params, body.node, isAsync)
      )
    }
    const classProp = t.classProperty(
      t.identifier(name),
      value.isFunctionExpression() || value.isArrowFunctionExpression()
        ? t.arrowFunctionExpression(value.node.params, value.node.body, isAsync)
        : value.node
    ) as any

    if (staticProps.includes(name)) {
      classProp.static = true
    }

    return classProp
  })

  if (defaultProps.length) {
    let classProp = t.classProperty(t.identifier('defaultProps'), t.objectExpression(
      defaultProps.map(p => t.objectProperty(t.identifier(p.name), p.value))
    )) as any
    classProp.static = true
    classBody.unshift(classProp)
  }

  if (json && t.isObjectExpression(json)) {
    classBody.push(t.classProperty(t.identifier('config'), json))
  }

  if (componentType === 'App') {
    let hasWillMount = false
    const globalData = template(`this.$app.globalData = this.globalData`)()
    for (const method of classBody) {
      if (!method) {
        continue
      }
      if (!t.isClassMethod(method)) {
        continue
      }
      if (t.isIdentifier(method.key, { name: Lifecycle.componentWillMount })) {
        hasWillMount = true
        method.body.body.unshift(globalData)
      }
    }
    if (!hasWillMount) {
      classBody.push(
        t.classMethod(
          'method',
          t.identifier(Lifecycle.componentWillMount),
          [],
          t.blockStatement([globalData])
        )
      )
    }
  }

  const renderFunc = buildRender(returned, stateKeys, propsKeys)

  return t.classDeclaration(
    t.identifier(componentType === 'App' ? 'App' : defaultClassName),
    t.memberExpression(t.identifier('Taro'), t.identifier('Component')),
    t.classBody(
      classBody.concat(renderFunc)
    ),
    []
  )
}
