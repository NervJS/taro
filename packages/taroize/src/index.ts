import * as t from 'babel-types'
import { parseWXML } from './wxml'
import { parseScript } from './script'
import { parseJSON } from './json'
import { errors } from './global'

interface Option {
  json?: string,
  script?: string,
  wxml?: string,
  path: string
}

export function parse (option: Option) {
  const { wxml, wxses, imports, refIds } = parseWXML(option.path, option.wxml)
  const json = parseJSON(option.json)
  const ast = parseScript(option.script, wxml as t.Expression, json, wxses, refIds)
  return {
    ast,
    imports,
    errors
  }
}
