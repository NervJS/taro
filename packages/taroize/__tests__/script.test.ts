import * as t from '@babel/types'

import { parseScript } from '../src/script'
import { parseWXML, WXS } from '../src/wxml'


interface Option {
  script: string
  scriptPath: string
  wxml: object
  wxses: WXS[]
  refIds: Set<string>
  isApp: boolean
}

const option: Option = {
  script: '',
  scriptPath: '',
  wxml: {},
  wxses: [],
  refIds: new Set<string>(),
  isApp: false
}

describe('parseScript', () => {
  // 调用 parseWXML 需要用到的参数，因parseWXML会先获取缓存，所有每个用例的path需要保持其唯一性
  let path = ''
  // 解析app.js
  test('app.js', () => {
    option.script = `
      App({
          onLaunch(){
              console.log('onLaunch');
          }
      })
    `
    option.wxml = { type: 'NullLiteral' }
    const ast = parseScript(option.script, option.scriptPath, option.wxml as t.Expression, option.wxses, option.refIds, true)
    expect(ast).toBeTruthy()
    expect(ast).toMatchSnapshot()
  })

  // 当存在 getAPP 或 getCurrentPages 方法
  test('getAPP || getCurrentPages', () => {
    option.script = `
      App({
          onLaunch(){
              const appInstance = getApp();
              console.log('页面栈信息：', getCurrentPages());
          }
      })
    `
    option.wxml = { type: 'NullLiteral' }
    const ast = parseScript(option.script, option.scriptPath, option.wxml as t.Expression, option.wxses, option.refIds, true)
    expect(ast).toMatchSnapshot()
  })

  // 当wxml是纯文本 parseWXML的返回值 { wxml } 影响js转换
  test('wxml jsxText', () => {
    option.script = `App({})`
    const wxmlStr = `123`
    // 调用 parseWXML 会首先获取缓存，所以第一个参数path建议保持唯一
    path = 'wxml_jsxText'
    const { wxml } = parseWXML(path, wxmlStr)
    const ast = parseScript(option.script, option.scriptPath, wxml as t.Expression, option.wxses, option.refIds, true)
    expect(ast).toMatchSnapshot()
  })

  // 当wxml存在变量 parseWXML的返回值 { wxml } 影响js转换
  test('wxml expression', () => {
    option.script = `
      App({
        data:{
            srt:'Hello Word!',
        }
      })
    `
    const wxmlStr = `<view>{{str}}</view>`
    // 调用 parseWXML 会首先获取缓存，所以第一个参数path建议保持唯一
    path = 'wxml_expression'
    const { wxml } = parseWXML(path, wxmlStr)
    const ast = parseScript(option.script, option.scriptPath, wxml as t.Expression, option.wxses, option.refIds, true)
    expect(ast).toMatchSnapshot()
  })
})
