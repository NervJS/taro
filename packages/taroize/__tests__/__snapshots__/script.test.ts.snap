// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseScript app.js 1`] = `
"import \\"@tarojs/components\\";
import React from \\"react\\";
import Taro from \\"@tarojs/taro\\";
const {
  default: withWeapp
} = require(\\"@tarojs/with-weapp\\");
App({
  onLaunch() {
    console.log('onLaunch');
  }
});
@withWeapp(cacheOptions.getOptionsFromCache(), true)
class App extends React.Component {
  render() {
    return this.props.children;
  }
}
export default App;"
`;

exports[`parseScript getAPP || getCurrentPages 1`] = `
"import \\"@tarojs/components\\";
import React from \\"react\\";
import Taro from \\"@tarojs/taro\\";
const {
  default: withWeapp
} = require(\\"@tarojs/with-weapp\\");
App({
  onLaunch() {
    const appInstance = Taro.getApp();
    console.log('页面栈信息：', Taro.getCurrentPages());
  }
});
@withWeapp(cacheOptions.getOptionsFromCache(), true)
class App extends React.Component {
  render() {
    return this.props.children;
  }
}
export default App;"
`;

exports[`parseScript wxml expression 1`] = `
"import { Block, View } from \\"@tarojs/components\\";
import React from \\"react\\";
import Taro from \\"@tarojs/taro\\";
const {
  default: withWeapp
} = require(\\"@tarojs/with-weapp\\");
App({
  data: {
    srt: 'Hello Word!'
  }
});
@withWeapp(cacheOptions.getOptionsFromCache(), true)
class App extends React.Component {
  render() {
    return this.props.children;
  }
}
export default App;"
`;

exports[`parseScript wxml jsxText 1`] = `
"import { Block } from \\"@tarojs/components\\";
import React from \\"react\\";
import Taro from \\"@tarojs/taro\\";
const {
  default: withWeapp
} = require(\\"@tarojs/with-weapp\\");
App({});
@withWeapp(cacheOptions.getOptionsFromCache(), true)
class App extends React.Component {
  render() {
    return this.props.children;
  }
}
export default App;"
`;
